File: app/src/main/java/com/gandhasiri/app/TreeAdapter.java
package com.gandhasiri.app;

import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import com.gandhasiri.app.databinding.ItemTreeBinding;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class TreeAdapter extends RecyclerView.Adapter<TreeAdapter.TreeViewHolder> {
    private List<Tree> trees = new ArrayList<>();
    private List<Float> distances = new ArrayList<>();
    private OnTreeClickListener listener;

    public interface OnTreeClickListener {
        void onTreeClick(Tree tree);
    }

    public void setOnTreeClickListener(OnTreeClickListener listener) {
        this.listener = listener;
    }

    public void setTrees(List<Tree> trees) {
        this.trees = trees;
        this.distances.clear();
        notifyDataSetChanged();
    }
    
    public void setTreesWithDistances(List<Tree> trees, List<Float> distances) {
        this.trees = trees;
        this.distances = distances;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public TreeViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemTreeBinding binding = ItemTreeBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new TreeViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull TreeViewHolder holder, int position) {
        Tree tree = trees.get(position);
        holder.binding.tvTreeId.setText("ID: " + tree.getTreeId());
        holder.binding.tvTreeGirth.setText(String.format(Locale.getDefault(), "Girth: %.1f cm", tree.getGirth()));
        
        // Maturity logic
        double maturityPercent = MaturityCalculator.calculateMaturityPercentage(tree.getGirth(), tree.getDatePlanted());
        int yearsLeft = MaturityCalculator.estimateYearsToHarvest(tree.getGirth(), tree.getDatePlanted());
        
        String estimateText = String.format(Locale.getDefault(), "Maturity: %.0f%% (%d years left)", maturityPercent, yearsLeft);

        // Distance logic (for Map view)
        if (distances != null && position < distances.size()) {
            estimateText = String.format(Locale.getDefault(), "Dist: %.1f m | %s", distances.get(position), estimateText);
        }
        
        holder.binding.tvMaturityEstimate.setText(estimateText);
        holder.binding.pbMaturity.setProgress((int) maturityPercent);

        if (tree.getPhotoPath() != null && !tree.getPhotoPath().isEmpty()) {
            Glide.with(holder.itemView.getContext())
                    .load(Uri.parse(tree.getPhotoPath()))
                    .placeholder(android.R.drawable.ic_menu_gallery)
                    .into(holder.binding.ivTreeThumbnail);
        }

        holder.itemView.setOnClickListener(v -> {
            if (listener != null) listener.onTreeClick(tree);
        });
    }

    @Override
    public int getItemCount() {
        return trees.size();
    }

    static class TreeViewHolder extends RecyclerView.ViewHolder {
        ItemTreeBinding binding;
        TreeViewHolder(ItemTreeBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/gandhasiri/app/TreeAdapter.kt
package com.gandhasiri.app

import android.net.Uri
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.gandhasiri.app.databinding.ItemTreeBinding
import java.util.*

class TreeAdapter : RecyclerView.Adapter<TreeAdapter.TreeViewHolder>() {
    private var trees: List<Tree> = ArrayList()
    private var distances: List<Float>? = null
    private var listener: OnTreeClickListener? = null

    fun interface OnTreeClickListener {
        fun onTreeClick(tree: Tree)
    }

    fun setOnTreeClickListener(listener: OnTreeClickListener) {
        this.listener = listener
    }

    fun setTrees(trees: List<Tree>) {
        this.trees = trees
        this.distances = null
        notifyDataSetChanged()
    }
    
    fun setTreesWithDistances(trees: List<Tree>, distances: List<Float>) {
        this.trees = trees
        this.distances = distances
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TreeViewHolder {
        val binding = ItemTreeBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return TreeViewHolder(binding)
    }

    override fun onBindViewHolder(holder: TreeViewHolder, position: Int) {
        val tree = trees[position]
        holder.binding.tvTreeId.text = "ID: ${tree.treeId}"
        holder.binding.tvTreeGirth.text = String.format(Locale.getDefault(), "Girth: %.1f cm", tree.girth)
        
        // Maturity logic
        val maturityPercent = MaturityCalculator.calculateMaturityPercentage(tree.girth, tree.datePlanted)
        val yearsLeft = MaturityCalculator.estimateYearsToHarvest(tree.girth, tree.datePlanted)
        
        var estimateText = String.format(Locale.getDefault(), "Maturity: %.0f%% (%d years left)", maturityPercent, yearsLeft)

        // Distance logic (for Map view)
        distances?.let {
            if (position < it.size) {
                estimateText = String.format(Locale.getDefault(), "Dist: %.1f m | %s", it[position], estimateText)
            }
        }
        
        holder.binding.tvMaturityEstimate.text = estimateText
        holder.binding.pbMaturity.progress = maturityPercent.toInt()

        if (tree.photoPath.isNotEmpty()) {
            Glide.with(holder.itemView.context)
                .load(Uri.parse(tree.photoPath))
                .placeholder(android.R.drawable.ic_menu_gallery)
                .into(holder.binding.ivTreeThumbnail)
        }

        holder.itemView.setOnClickListener {
            listener?.onTreeClick(tree)
        }
    }

    override fun getItemCount(): Int = trees.size

    class TreeViewHolder(val binding: ItemTreeBinding) : RecyclerView.ViewHolder(binding.root)
}
